File: Protocol Bernardo/Engines/DataAcquisitionEngine/Extensions/PAEStatus.swift
//
//  DAEStatus.swift
//  Protocol Bernardo
//
//  Created by Valentin Dufois on 2019-01-22.
//  Copyright © 2019 Prisme. All rights reserved.
//

import Foundation

/// Represent the current state of the C++ DAE
///
/// You should not use this structure directly as it is converted to `ConnectedDevices` upon
/// receiving.
extension PAEStatus {
    /// List of all the devices in a Swift friendly format
    func copyAndDeallocate() -> [Serial: PhysicalDevice] {
        // The map
        var devicesMap = [Serial: PhysicalDevice]()

        // Pointer used for parsing
        var pointer = connectedDevices
        
        // Loop on each device
        for i in 0..<deviceCount {
            let daeDeviceStatus = pointer!.pointee
            
            // Insert in the map
            devicesMap[daeDeviceStatus.serial] = PhysicalDevice(from: daeDeviceStatus)
            
            // Free it
            daeDeviceStatus.trackedUsers.deallocate()
            
            // Check if we can advance
            if i + 1 < deviceCount {
                pointer = pointer?.successor()
            }
        }
        
        // Free the array
        connectedDevices.deallocate()
        
        return devicesMap
    }
}

File: PositionAcquisitionEngine/Structs/PAEStatus.cpp
//
//  PAEStatus.cpp
//  PositionAcquisitionEngine
//
//  Created by Valentin Dufois on 2019-04-02.
//  Copyright © 2019 Prisme. All rights reserved.
//

#include "PAEStatus.h"

PAEStatus * PAEStatus_copy(PAEStatus * s)
{
	PAEStatus * c = new PAEStatus();

	strcpy(c->hostname, s->hostname);
	c->deviceCount = s->deviceCount;
	c->connectedDevices = (PAEDeviceStatus *)malloc(sizeof(PAEDeviceStatus) * c->deviceCount);

	for(int i = 0; i < c->deviceCount; ++i) {
		c->connectedDevices[i] = PAEDeviceStatus_copy(&(s->connectedDevices[i]));
	}

	return c;
}
